File: demo/GazeboMonitor/MapViewer.py
'''
Created on Apr 22, 2014

@author: walter
'''

from PyQt4 import QtGui, QtCore
import math

class MapViewer(QtGui.QLabel):
    '''
    classdocs
    '''

    def __init__(self, parent):
        super(MapViewer, self).__init__(parent)
        self.mPoint = QtCore.QPoint(5,5)
        self.mPointYaw = 0.0;
        self.mShowPoint = False
        self.plannedPath = None
        self.plannedPathReachedIdx = 0
    
        
    def paintEvent(self, e):
        
        super(MapViewer, self).paintEvent(e)
        
        if self.mShowPoint == True:
            qp = QtGui.QPainter()
            qp.begin(self)
            qp.setPen(QtGui.QPen(QtGui.QColor(255,0,0)))
            qp.setBrush(QtGui.QBrush(QtGui.QColor(255,0,0)))
            qp.drawEllipse(self.mPoint, 5, 5)
            orientX = self.mPoint.x() + 10.0 * math.cos(self.mPointYaw)
            orientY = self.mPoint.y() + 10.0 * math.sin(self.mPointYaw)
            qp.setPen(QtGui.QPen(QtGui.QColor(0,0,255)))
            qp.drawLine(self.mPoint, QtCore.QPoint(orientX, orientY))
            
            if self.plannedPath != None:
                
                for t in range(self.plannedPath.length):
                    pt = self.plannedPath.waypoints[t]
                    if t < self.plannedPathReachedIdx:
                        qp.setPen(QtGui.QPen(QtGui.QColor(0,0,255)))
                        qp.setBrush(QtGui.QBrush(QtGui.QColor(0,0,255)))
                    else:
                        qp.setPen(QtGui.QPen(QtGui.QColor(0,255,0)))
                        qp.setBrush(QtGui.QBrush(QtGui.QColor(0,255,0)))
                    qp.drawEllipse(QtCore.QPoint(int(pt[0]), int(pt[1])), 3, 3)
                    
                qp.setPen(QtGui.QPen(QtGui.QColor(0,0,0), 1, QtCore.Qt.DashLine))
                
                if self.plannedPathReachedIdx < self.plannedPath.length:
                    targetPoint = self.plannedPath.waypoints[self.plannedPathReachedIdx]
                    targetX = int(targetPoint[0])
                    targetY = int(targetPoint[1])                
                    qp.drawLine(self.mPoint, QtCore.QPoint(targetX, targetY))
            
            qp.end()
            
    def setPoint(self,x,y):
        self.mPoint.setX(x)
        self.mPoint.setY(y)
        
    
            
        

File: demo/RoboStageMonitor/mapviewer.cpp
#include "mapviewer.h"
#include <QPainter>
#include <QtCore/qmath.h>

MapViewer::MapViewer(QWidget *parent) :
    QLabel(parent)
{
   mPoint = new QPoint();
   mShowPoint = false;
   mpPath = new DiscretePath();
}


void MapViewer::paintEvent(QPaintEvent * e)
{
    QLabel::paintEvent(e);

    if(mShowPoint)
    {
        QPainter painter(this);
        painter.setPen(QPen(Qt::red));
        painter.setBrush(QBrush(Qt::red, Qt::SolidPattern));
        painter.drawEllipse(*mPoint, 5 ,5);
        qDebug("paint x %d y %d", mPoint->x(), mPoint->y());
        double orientX = mPoint->x() + 10.0 * qCos(mPointYaw * 3.141592/180.0);
        double orientY = mPoint->y() + 10.0 * qSin(mPointYaw * 3.141592/180.0);
        QPointF orientPoint(orientX, orientY);
        painter.setPen(QPen(Qt::blue));
        painter.drawLine(*mPoint, orientPoint);
    }

    if(mpPath)
    {
        for(int i=0;i< mpPath->length();i++)
        {
            QPointF pointf = mpPath->mSequence[i];
            QPoint point((int)pointf.x(), (int)pointf.y());
            QPainter painter(this);
            painter.setPen(QPen(Qt::green));
            painter.setBrush(QBrush(Qt::green, Qt::SolidPattern));
            painter.drawEllipse(point, 2 ,2);
            //qDebug("point x %d y %d", point.x(), point.y());
        }
    }

}

void MapViewer::setPoint(int x, int y)
{
    if(mPoint)
    {
        mPoint->setX(x);
        mPoint->setY(y);
    }
}

void MapViewer::showPoint(bool show)
{
    mShowPoint = show;
}

void MapViewer::loadPath(DiscretePath path)
{
    *mpPath = path;
}

void MapViewer::setPointYaw(double yaw)
{
    mPointYaw = yaw;
}
